File: iteration2/src/LoginInterface.java
import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;

import java.util.ArrayList;
import java.util.Scanner;
import java.io.FileReader;
import java.io.IOException;

public class LoginInterface {

    // Attributes
    private Session session;
    private Scanner scanner = new Scanner(System.in);

    // Methods

    // login method to get the user in
    public void login() {
        while (true) {
            System.out.println(Colors.RED
                    + "\n--------------------Marmara Course Registration System--------------------\n" + Colors.RESET);
            System.out.println("Please enter your ID and password to login!\n");
            // get user login info
            String ID = getUserID();
            String password = getUserPassword();
            // check user type
            if (getUserType(ID).equalsIgnoreCase("student")) {
                // check if user exists
                if (userExists("students", ID)) {
                    if (checkUserLoginInfo("students", ID, password)) {
                        // create student object
                        session = new Session(createStudent(ID, null));
                        System.out.println("\nWelcome " + session.getUser().getName() + " "
                                + session.getUser().getSurname() + "!");
                        System.out.println("You have " + Colors.GREEN + "successfully" + Colors.RESET + " logged in.");
                        // direct student to student menu
                        System.out.println("You will be directed to the Student Menu!\n");
                        StudentInterface studentInterface = new StudentInterface(session, this);
                        studentInterface.stuMenu();
                    } else {
                        System.out.println(Colors.YELLOW + "Invalid ID or password! Please try again." + Colors.RESET);
                        continue;
                    }
                } else {
                    System.out.println(Colors.YELLOW + "Invalid ID! Please try again." + Colors.RESET);
                    continue;
                }
            } else if (getUserType(ID).equalsIgnoreCase("lecturer")) {
                // check if user exists
                if (userExists("lecturers", ID)) {
                    if (checkUserLoginInfo("advisors", ID, password)) {
                        // create advisor object
                        session = new Session(createAdvisor(ID, "null"));
                        System.out.println("\nWelcome " + session.getUser().getName() + " "
                                + session.getUser().getSurname() + "!");
                        System.out.println("You have " + Colors.GREEN + "successfully" + Colors.RESET + " logged in.");
                        System.out.println("You will be directed to the Advisor Menu!\n");
                        // direct advisor to advisor menu
                        AdvisorInterface advisorInterface = new AdvisorInterface(session, this);
                        advisorInterface.advMenu();
                    } else if (checkUserLoginInfo("lecturers", ID, password)) {
                        // create lecturer object
                        session = new Session(createLecturer(ID));
                        System.out.println("\nWelcome " + session.getUser().getName() + " "
                                + session.getUser().getSurname() + "!");
                        System.out.println("You have " + Colors.GREEN + "successfully" + Colors.RESET + " logged in.");
                        System.out.println("You will be directed to the Lecturer Menu!\n");
                        // direct lecturer to lecturer menu
                        AdvisorInterface advisorInterface = new AdvisorInterface(session, this);
                        advisorInterface.advMenu();
                    } else {
                        System.out.println(Colors.YELLOW + "Invalid ID or password! Please try again." + Colors.RESET);
                        continue;
                    }
                } else {
                    System.out.println(Colors.YELLOW + "Invalid ID! Please try again." + Colors.RESET);
                    continue;
                }
            } else {
                System.out.println(Colors.YELLOW + "Invalid ID or password! Please try again." + Colors.RESET);
                continue;
            }
        }

    }

    public void logout() {
        while (true) {
            // if user wants to logout and exit
            System.out.println("Do you want to logout and exit? (" + Colors.YELLOW + "y/n" + Colors.RESET + ")");
            String answer = scanner.nextLine();
            if (answer.equals("y")) {
                System.out.println(
                        "You have " + Colors.GREEN + "successfully" + Colors.RESET + " logged out and exited.");
                System.out.println(
                        Colors.RED
                                + "\n<<<--------Thank you for using Marmara Course Registration System-------->>>\n"
                                + Colors.RESET);
                // set person to null to recreate it when logging in again
                session.setUser(null);
                exit();
            } else if (answer.equals("n")) {
                System.out.println(Colors.YELLOW + "You will be redirected to the login menu." + Colors.RESET);
                // set person to null to recreate it when logging in again
                session.setUser(null);
                login();
            } else {
                System.out.println(Colors.YELLOW + "Invalid input! Please try again." + Colors.RESET);
                continue;
            }
        }
    }

    // shuw down the program
    public void exit() {
        System.exit(0);
    }

    private String getUserID() {
        // get id
        while (true) {
            System.out.print(Colors.YELLOW + "ID: " + Colors.RESET);
            String ID = scanner.nextLine();
            // check id format
            if (!idFormatChecker(ID)) {
                System.out.println(Colors.YELLOW + "Invalid input! Please try again." + Colors.RESET);
            } else {
                return ID;
            }
        }

    }

    private String getUserPassword() {
        // get password
        while (true) {
            System.out.print(Colors.YELLOW + "Password: " + Colors.RESET);
            String password = scanner.nextLine();
            if (!passwordFormatChecker(password)) {
                System.out.println(Colors.YELLOW + "Invalid input! Please try again." + Colors.RESET);
            } else {
                return password;
            }
        }
    }

    // returns true if user with the given id exists
    private boolean userExists(String fileName, String ID) {
        JSONParser parser = new JSONParser();
        try {
            FileReader reader = new FileReader("jsons/" + fileName + ".json");
            Object obj = parser.parse(reader);
            JSONArray userList = (JSONArray) obj;

            for (Object userObject : userList) {
                JSONObject user = (JSONObject) userObject;
                if (user.get("ID").equals(ID)) {
                    return true;
                }
            }
        } catch (IOException | ParseException e) {
            e.printStackTrace();
        }
        return false;
    }

    // returns true if user with the given id and password exists
    private boolean checkUserLoginInfo(String fileName, String ID, String password) {
        while (true) {
            JSONParser parser = new JSONParser();
            try {
                FileReader reader = new FileReader("jsons/" + fileName + ".json");
                Object obj = parser.parse(reader);
                JSONArray userList = (JSONArray) obj;

                for (Object userObject : userList) {
                    JSONObject user = (JSONObject) userObject;
                    if (user.get("ID").equals(ID) && user.get("Password").equals(password)) {
                        return true;
                    }
                }
            } catch (IOException | ParseException e) {
                e.printStackTrace();
            }
            return false;
        }

    }

    // creates student object
    private Student createStudent(String ID, Advisor advisor) {
        Object studentObj = null;
        try {
            studentObj = new JSONParser().parse(new FileReader("jsons/student/" + ID + ".json"));
        } catch (IOException | ParseException e) {
            e.printStackTrace();
        }
        JSONObject student = (JSONObject) studentObj;

        // user attributes in json
        String name = (String) student.get("Name");
        String surname = (String) student.get("Surname");
        String email = (String) student.get("EMail");
        String phoneNumber = (String) student.get("PhoneNumber");
        String department = (String) student.get("Department");
        String advisorID = (String) student.get("AdvisorID");
        long semester = (long) student.get("Semester");
        String faculty = (String) student.get("Faculty");
        String studentID = (String) student.get("ID");
        String password = (String) student.get("Password");
        JSONArray takenCoursesArray = (JSONArray) student.get("TakenCourses");

        // if there is no advisor, create one
        if (advisor == null) {
            advisor = createAdvisor(advisorID, studentID);
        }

        Student stu = new Student(name, surname, email, phoneNumber, studentID, password, faculty, department,
                (int) semester, advisor);

        // if student has no taken courses, skip this part
        if (takenCoursesArray.isEmpty()) {
            // do nothing
        } else {
            // Create a course object with the given course id's
            for (Object courseObj : takenCoursesArray) {
                JSONObject course = (JSONObject) courseObj;
                String courseID = (String) course.get("ID");
                // add the course to student's taken courses list
                stu.setCurrentTakenCourses(createCourse(courseID));
            }
        }

        return stu;
    }

    // creates advisor object
    private Advisor createAdvisor(String advisorID, String studentId) {
        JSONParser parser = new JSONParser();
        try {
            FileReader reader = new FileReader("jsons/advisors.json");
            Object obj = parser.parse(reader);
            JSONArray advisorList = (JSONArray) obj;

            for (Object advisorObj : advisorList) {
                JSONObject advisor = (JSONObject) advisorObj;
                if (advisor.get("ID").equals(advisorID)) {
                    // user attributes in json
                    String name = (String) advisor.get("Name");
                    String surname = (String) advisor.get("Surname");
                    String email = (String) advisor.get("EMail");
                    String phoneNumber = (String) advisor.get("PhoneNumber");
                    String faculty = (String) advisor.get("Faculty");
                    String department = (String) advisor.get("Department");
                    String password = (String) advisor.get("Password");
                    String academicTitle = (String) advisor.get("AcademicTitle");
                    JSONArray studentList = (JSONArray) advisor.get("Students");
                    JSONArray givenCourses = (JSONArray) advisor.get("Courses");
                    Advisor adv = new Advisor(name, surname, email, phoneNumber, advisorID, password, faculty,
                            department, academicTitle);
                    // add students to advisor's student list
                    for (Object studentObj : studentList) {
                        JSONObject student = (JSONObject) studentObj;
                        String studentId2 = (String) student.get("ID");
                        // if the students exists in the advisor's student list, skip it
                        if (studentId.equals(studentId2)) {
                            continue;
                        }
                        adv.setStudent(createStudent(studentId2, adv));
                    }
                    // add courses to advisor's course list
                    for (Object courseObj : givenCourses) {
                        String courseCode = (String) courseObj;
                        adv.setCourse(createCourse(courseCode));
                    }
                    return adv;
                }
            }
        } catch (IOException | ParseException e) {
            e.printStackTrace();
        }
        return null; // return null if no advisor with the given ID is found
    }

    // create lecturer object
    private Lecturer createLecturer(String lecturerID) {
        JSONParser parser = new JSONParser();
        try {
            FileReader reader = new FileReader("jsons/lecturers.json");
            Object obj = parser.parse(reader);
            JSONArray lecturerList = (JSONArray) obj;

            for (Object lecturerObj : lecturerList) {
                JSONObject lecturer = (JSONObject) lecturerObj;
                if (lecturer.get("ID").equals(lecturerID)) {
                    // user attributes in json
                    String name = (String) lecturer.get("Name");
                    String surname = (String) lecturer.get("Surname");
                    String email = (String) lecturer.get("EMail");
                    String phoneNumber = (String) lecturer.get("PhoneNumber");
                    String faculty = (String) lecturer.get("Faculty");
                    String department = (String) lecturer.get("Department");
                    String password = (String) lecturer.get("Password");
                    JSONArray givenCourses = (JSONArray) lecturer.get("Courses");
                    String academicTitle = (String) lecturer.get("AcademicTitle");
                    Lecturer lec = new Lecturer(name, surname, email, phoneNumber, lecturerID, password, faculty,
                            department, academicTitle);

                    // add courses to lecturer's course list
                    for (Object courseObj : givenCourses) {
                        String courseCode = (String) courseObj;

                        lec.setCourse(createCourse(courseCode));
                    }

                    return lec;

                }
            }
        } catch (IOException | ParseException e) {
            e.printStackTrace();
        }
        return null; // return null if no lecturer with the given ID is found
    }

    // create course object
    private Course createCourse(String courseCode) {
        JSONParser parser = new JSONParser();
        try {
            FileReader reader = new FileReader("jsons/CoursesOffered.json");
            Object obj = parser.parse(reader);
            JSONArray coursesList = (JSONArray) obj;

            for (Object courseObj : coursesList) {
                JSONObject course = (JSONObject) courseObj;
                if (course.get("CourseID").equals(courseCode)) {
                    // course attributes in json
                    String courseName = (String) course.get("CourseName");
                    String courseDayTimeLocation = (String) course.get("CourseDayTimeLocation");
                    CourseSession courseSession = createCourseSession(courseDayTimeLocation);
                    // decide whether it is lecture or lab
                    int dotCount = 0;
                    for (int i = 0; i < courseCode.length(); i++) {
                        if (courseCode.charAt(i) == '.') {
                            dotCount++;
                        }
                    }
                    if (dotCount == 1) {
                        // create lecture
                        Lecture lec = new Lecture(courseName, courseCode, courseSession);
                        return lec;
                    } else if (dotCount == 2) {
                        // create lab
                        Lab lab = new Lab(courseName, courseCode, courseSession);
                        return lab;
                    } else if (dotCount == 0) {
                        // create lab
                        Course cou = new Course(courseName, courseCode, courseSession);
                        return cou;
                    } else {
                        System.out.println(Colors.CYAN + "Invalid course code!" + Colors.RESET);
                        continue;
                    }
                }
            }
        } catch (IOException | ParseException e) {
            e.printStackTrace();
        }
        return null; // return null if no course with the given code is found
    }

    // create course session object
    private CourseSession createCourseSession(String courseDayTimeLocation) {
        // parse the string and create course session object
        String[] parts = courseDayTimeLocation.split(" ");
        ArrayList<String> courseDay = new ArrayList<>();
        ArrayList<String> courseStartTime = new ArrayList<>();
        ArrayList<String> courseEndTime = new ArrayList<>();
        ArrayList<String> coursePlace = new ArrayList<>();
        // each object of the list is a different course session
        for (int i = 0; i < parts.length; i += 5) {
            courseDay.add(parts[i]);
            courseStartTime.add(parts[i + 1]);
            courseEndTime.add(parts[i + 3]);
            coursePlace.add(parts[i + 4]);
        }

        return new CourseSession(courseDay, courseStartTime, courseEndTime, coursePlace);
    }

    // returns false if id format is wrong
    private boolean idFormatChecker(String ID) {
        if (ID.isEmpty()) {
            return false;
        }
        for (int i = 0; i < ID.length(); i++) {
            if (ID.charAt(i) < '0' || ID.charAt(i) > '9') {
                return false;
            }
        }
        if (ID.length() != 9 && ID.length() != 6) {
            return false;
        }
        return true;
    }

    // returns false if password format is wrong
    private boolean passwordFormatChecker(String password) {
        if (password.isEmpty()) {
            return false;
        }
        if (password.contains(" ")) {
            return false;
        }
        return true;
    }

    // returns user type
    private String getUserType(String ID) {
        if (ID.length() == 9) {
            return "student";
        } else if (ID.length() == 6) {
            return "lecturer";
        } else {
            return null;
        }
    }
}

File: iteration3/src/login_interface.py
from datetime import datetime
import json
import sys
import logging
import os

from student_interface import StudentInterface
from advisor_interface import AdvisorInterface
from course import Course
from lecture import Lecture
from lab import Lab
from course_session import CourseSession
from advisor import Advisor
from student import Student
from lecturer import Lecturer
from colors import Colors
from session import Session


class LoginInterface:
    logging.basicConfig(filename='./logs/app.log', filemode='w', format='%(name)s - %(levelname)s - %(message)s', level=logging.INFO)
    def __init__(self):
        self.__colors = Colors()
        self.__session: Session = None

    def login(self):
        while True:
            print(
                self.__colors.get_bold()
                + self.__colors.get_red()
                + "\nMarmara Course Registration System\n"
                + self.__colors.get_reset()
                + self.__colors.get_reset()
            )
            print("Please enter your ID and password to login!\n")
            id = self.__get_user_id()
            password = self.__get_user_password()

            user_type = self.__get_user_type(id)
            if user_type == "student":
                if self.__user_exists("students", id):
                    if self.__check_user_login_info("students", id, password):
                        self.__session = Session(self.__create_student(id, None))
                        print(
                            "\nWelcome {} {}!".format(
                                self.__session.get_user().get_name(),
                                self.__session.get_user().get_surname(),
                            )
                        )
                        print(
                            "You have {} successfully {} logged in.".format(
                                self.__colors.get_green(), self.__colors.get_reset()
                            )
                        )
                        logging.info("time: {}, User {} logged in.".format(datetime.now().strftime("%Y-%m-%d %H:%M:%S"), self.__session.get_user().get_id()))
                        print("You will be directed to the Student Menu!\n")
                        student_interface = StudentInterface(self.__session, self)
                        student_interface.stu_menu()
                    else:
                        print(
                            "{}Invalid ID or password! Please try again.{}".format(
                                self.__colors.get_yellow(), self.__colors.get_reset()
                            )
                        )
                        logging.warning("time: {}, User {} invalid password.".format(datetime.now().strftime("%Y-%m-%d %H:%M:%S"), id))
                        continue
                else:
                    print(
                        "{}Invalid ID! Please try again.{}".format(
                            self.__colors.get_yellow(), self.__colors.get_reset()
                        )
                    )
                    continue
            elif user_type == "lecturer":
                if self.__user_exists("lecturers", id):
                    if self.__check_user_login_info("advisors", id, password):
                        self.__session = Session(self.__create_advisor(id, "null"))
                        print(
                            "\nWelcome {} {}!".format(
                                self.__session.get_user().get_name(),
                                self.__session.get_user().get_surname(),
                            )
                        )
                        print(
                            "You have {} successfully {} logged in.".format(
                                self.__colors.get_green(), self.__colors.get_reset()
                            )
                        )
                        logging.info("time: {}, User {} logged in.".format(datetime.now().strftime("%Y-%m-%d %H:%M:%S"), self.__session.get_user().get_id()))
                        print("You will be directed to the Advisor Menu!\n")
                        advisor_interface = AdvisorInterface(self.__session, self)
                        advisor_interface.adv_menu()
                    elif self.__check_user_login_info("lecturers", id, password):
                        self.__session = Session(self.__create_lecturer(id))
                        print(
                            "\nWelcome {} {}!".format(
                                self.__session.get_user().get_name(),
                                self.__session.get_user().get_surname(),
                            )
                        )
                        print(
                            "You have {} successfully {} logged in.".format(
                                self.__colors.get_green(), self.__colors.get_reset()
                            )
                        )
                        logging.info("time: {}, User {} logged in.".format(datetime.now().strftime("%Y-%m-%d %H:%M:%S"), self.__session.get_user().get_id()))
                        print("You will be directed to the Lecturer Menu!\n")
                        advisor_interface = AdvisorInterface(self.__session, self)
                        advisor_interface.adv_menu()
                    else:
                        print(
                            "{}Invalid ID or password! Please try again.{}".format(
                                self.__colors.get_yellow(), self.__colors.get_reset()
                            )
                        )
                        logging.warning("time: {}, User {} invalid password.".format(datetime.now().strftime("%Y-%m-%d %H:%M:%S"), id))
                        continue
                else:
                    print(
                        "{}Invalid ID! Please try again.{}".format(
                            self.__colors.get_yellow(), self.__colors.get_reset()
                        )
                    )
                    continue
            else:
                print(
                    "{}Invalid ID or password! Please try again.{}".format(
                        self.__colors.get_yellow(), self.__colors.get_reset()
                    )
                )
                continue

    def logout(self):
        while True:
            print(
                "Do you want to logout and exit? ({}/{})".format(
                    self.__colors.get_yellow() + "y" + self.__colors.get_reset(),
                    self.__colors.get_yellow() + "n" + self.__colors.get_reset(),
                )
            )
            answer = input()
            if answer == "y":
                print(
                    "You have {} successfully {} logged out and exited.".format(
                        self.__colors.get_green(), self.__colors.get_reset()
                    )
                )
                logging.info("time: {}, User {} logged out.".format(datetime.now().strftime("%Y-%m-%d %H:%M:%S"), self.__session.get_user().get_id()))
                print(
                    "{}{}Thank you for using Marmara Course Registration System{}".format(
                        self.__colors.get_red(),
                        self.__colors.get_bold(),
                        self.__colors.get_reset(),
                    )
                )
                self.__session.set_user(None)
                self.exit()
            elif answer == "n":
                print(
                    "{}You will be redirected to the login menu.{}".format(
                        self.__colors.get_yellow(), self.__colors.get_reset()
                    )
                )
                self.__session.set_user(None)
                self.login()
            else:
                print(
                    "{}Invalid input! Please try again.{}".format(
                        self.__colors.get_yellow(), self.__colors.get_reset()
                    )
                )
                continue

    def exit(self):
        sys.exit(0)

    def __get_user_id(self):
        while True:
            print(
                "{}ID: {}".format(
                    self.__colors.get_yellow(), self.__colors.get_reset()
                ),
                end="",
            )
            id = input()
            if not self.__id_format_checker(id):
                print(
                    "{}Invalid input! Please try again.{}".format(
                        self.__colors.get_yellow(), self.__colors.get_reset()
                    )
                )
            else:
                return id

    def __get_user_password(self):
        while True:
            print(
                "{}Password: {}".format(
                    self.__colors.get_yellow(), self.__colors.get_reset()
                ),
                end="",
            )
            password = input()
            if not self.__password_format_checker(password):
                print(
                    "{}Invalid input! Please try again.{}".format(
                        self.__colors.get_yellow(), self.__colors.get_reset()
                    )
                )
            else:
                return password

    def __user_exists(self, file_name, id):
        try:
            with open("jsons/{}.json".format(file_name), "r", encoding='utf-8') as reader:
                user_list = json.load(reader)

                for user in user_list:
                    if user["ID"] == id:
                        return True
        except (IOError, json.JSONDecodeError):
            pass
        return False

    def __check_user_login_info(self, file_name, id, password):
        while True:
            try:
                with open("jsons/{}.json".format(file_name), "r", encoding='utf-8') as reader:
                    user_list = json.load(reader)

                    for user in user_list:
                        if user["ID"] == id and user["Password"] == password:
                            return True
            except (IOError, json.JSONDecodeError):
                pass
            return False

    def __create_student(self, id, advisor):
        try:
            with open("jsons/student/{}.json".format(id), "r") as reader:
                student = json.load(reader)

                name = student["Name"]
                surname = student["Surname"]
                email = student["EMail"]
                phone_number = student["PhoneNumber"]
                department = student["Department"]
                advisor_id = student["AdvisorID"]
                semester = student["Semester"]
                faculty = student["Faculty"]
                student_id = student["ID"]
                password = student["Password"]
                taken_courses_array = student["TakenCourses"]

                if advisor is None:
                    advisor = self.__create_advisor(advisor_id, student_id)

                stu = Student(
                    name,
                    surname,
                    email,
                    phone_number,
                    student_id,
                    password,
                    faculty,
                    department,
                    semester,
                    advisor,
                )

                if taken_courses_array:
                    for course_id in taken_courses_array:
                        stu.set_current_taken_courses(self.__create_course(course_id))

                return stu
        except (IOError, json.JSONDecodeError):
            pass
        return None

    def __create_advisor(self, advisor_id, student_id):
        try:
            with open("jsons/advisors.json", "r") as reader:
                advisor_list: list = json.load(reader)

                for advisor in advisor_list:
                    if advisor["ID"] == advisor_id:
                        name = advisor["Name"]
                        surname = advisor["Surname"]
                        email = advisor["EMail"]
                        phone_number = advisor["PhoneNumber"]
                        faculty = advisor["Faculty"]
                        department = advisor["Department"]
                        password = advisor["Password"]
                        academic_title = advisor["AcademicTitle"]
                        student_list = advisor["Students"]
                        given_courses = advisor["Courses"]

                        adv = Advisor(
                            name,
                            surname,
                            email,
                            phone_number,
                            advisor_id,
                            password,
                            faculty,
                            department,
                            academic_title,
                        )

                        for student in student_list:
                            student_id_2 = student["ID"]
                            if student_id == student_id_2:
                                continue
                            adv.set_student(self.__create_student(student_id_2, adv))

                        for course_code in given_courses:
                            adv.set_course(self.__create_course(course_code))

                        return adv
        except (IOError, json.JSONDecodeError):
            pass
        return None

    def __create_lecturer(self, lecturer_id):
        try:
            with open("jsons/lecturers.json", "r") as reader:
                lecturer_list: list = json.load(reader)

                for lecturer in lecturer_list:
                    if lecturer["ID"] == lecturer_id:
                        name = lecturer["Name"]
                        surname = lecturer["Surname"]
                        email = lecturer["EMail"]
                        phone_number = lecturer["PhoneNumber"]
                        faculty = lecturer["Faculty"]
                        department = lecturer["Department"]
                        password = lecturer["Password"]
                        given_courses = lecturer["Courses"]
                        academic_title = lecturer["AcademicTitle"]

                        lec = Lecturer(
                            name,
                            surname,
                            email,
                            phone_number,
                            lecturer_id,
                            password,
                            faculty,
                            department,
                            given_courses,
                            academic_title,
                        )

                        for course_code in given_courses:
                            lec.set_course(self.__create_course(course_code))

                        return lec
        except (IOError, json.JSONDecodeError):
            pass
        return None

    def __create_course(self, course_code: str):
        try:
            with open("jsons/CoursesOffered.json", "r", encoding='utf-8') as reader:
                courses_list = json.load(reader)

                for course_obj in courses_list:
                    if course_obj["CourseID"] == course_code:
                        course_name = course_obj["CourseName"]
                        course_day_time_location = course_obj["CourseDayTimeLocation"]
                        course_session = self.__create_course_session(
                            course_day_time_location
                        )

                        dot_count = course_code.count(".")
                        if dot_count == 1:
                            return Lecture(
                                course_name=course_name,
                                lecture_id=course_code,
                                course_session=course_session,
                            )
                        elif dot_count == 2:
                            return Lab(
                                course_name=course_name,
                                lab_id=course_code,
                                course_session=course_session,
                            )
                        elif dot_count == 0:
                            return Course(
                                course_name=course_name,
                                course_id=course_code,
                                course_session=course_session,
                            )
                        else:
                            print(
                                "{}Invalid course code!{}".format(
                                    self.__colors.get_cyan(), self.__colors.get_reset()
                                )
                            )
                            continue
        except (IOError, json.JSONDecodeError):
            pass
        return None

    def __create_course_session(self, course_day_time_location: str) -> CourseSession:
        parts = course_day_time_location.split(" ")
        course_day = parts[::5]
        course_start_time = parts[1::5]
        course_end_time = parts[3::5]
        course_place = parts[4::5]

        return CourseSession(
            course_day, course_start_time, course_end_time, course_place
        )

    def __id_format_checker(self, id: str):
        if not id or not id.isdigit():
            return False
        if len(id) != 9 and len(id) != 6:
            return False
        return True

    def __password_format_checker(self, password):
        if not password or " " in password:
            return False
        return True

    def __get_user_type(self, id):
        if len(id) == 9:
            return "student"
        elif len(id) == 6:
            return "lecturer"
        else:
            return None

